### DISCUSSION TITLE
Given a list of cities visit one in each state ...

### DISCUSSION THREAD

**Discussion by ra-esmith** [google/or-tools#3408] [Routing (and legacy CP) questions]
Hello,

I am currently stumped by how with TSP and Or Tools I might approach this problem.

I have a list of US cities and I wish to visit one per state, and minimize distance.  I want to further limit the number of cities to say 20 out of 52.  

I can use penalties (to limit to 20) and minimize distance (every example), that is fine.  But how would I structure a solution on OR Tools so that the cities in each state are a disjoint set ... such that I visit one and only one, and then move on ...

Thanks for the HINT!  
or-tools is crazy amazing, though a bit hard to wrap my mind around some problems like this.

Thanks !
Evan

**Comment by ra-esmith** [comment]
An idea ...?

I could have a dimension for every state in the USA
and then have a 
index = routing.End(vehicle_id)
dimension.CumulVar(index).SetMax(1)

But if given say 100 cities (all states) and I wish to visit 20. 
Would the penalties code which drops cities
Be able to deal with this SetMax which basically would say visit each state at most once ... or must be once.  And then not find a solution.

I am working in python.

Evan

**Comment by ra-esmith** [comment]
Ok the post https://github.com/google/or-tools/discussions/3327

Suggested something that might work here, making a disjoint set of the cities in each state, and having a cardinality of one.

However, (and this is where I am new to or tools).  The post uses a "solver".   So far I have been using examples from the doc page
Which are 
    solution = routing.SolveWithParameters(search_parameters)

So I am not sure how I can add a "constraint" using a solver.  
How does one use a solver with the examples in the documentation (I bet this is really clear to many folks ... sorry).  Still with that information I can likely move forward.

Thanks!
Evan

**Comment by ra-esmith** [comment]
Ok digging through documentation (C++ seems rich here)
https://developers.google.com/optimization/reference/constraint_solver/routing/RoutingModel
The discussion on
Adds a disjunction constraint on the indices: exactly 'max_cardinality' of the indices are active. Start and end indices of any vehicle cannot be part of a disjunction. If a penalty is given, at most 'max_cardinality' of the indices can be active, and if less are active, 'penalty' is payed per inactive index. This is equivalent to adding the constraint: p + Sum(i)active[i] == max_cardinality where p is an integer variable, and the following cost to the cost function: p * penalty. 'penalty' must be positive to make the disjunction optional; a negative penalty will force 'max_cardinality' indices of the disjunction to be performed, and therefore p == 0. Note: passing a vector with a single index will model an optional index with a penalty cost if it is not visited.

Might work for me.  I am thinking that I might be able to make a disjunction out of the cities in each state etc ... 

Fun
Evan

**Comment by ra-esmith** [comment]
Ok looks like on the router I can request the solver and add constraints!  that seems like what I must do.

and I guess is what is implied when folks say "use the solver" :-)


solver | Return type: Solver*Returns the underlying constraint solver. Can be used to add extra constraints and/or modify search algoithms.
-- | --


[solver](https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/routing.h#L1332)	
Return type: Solver*

Returns the underlying constraint solver. Can be used to add extra constraints and/or modify search algoithms.

**Reply by jmarca** [reply]
Sorry I was replying to the email first and only read the first message. If you want 20 cities max and one city per state max then add one more dimension, a “counting” dimension. Every city gives 1 for its state, 1 for the count dimension, and 0 for other states. Vehicle capacity is 1 in every state dimension, and 20 in the count dimension. 

I don’t see a need to use `routing.solver()` and all that, but I haven’t actually tried to do anything either.

**Reply by ra-esmith** [reply]
Hello,

First, Thanks for taking the time to think about this!
I am not sure yet what you are thinking.
Sounds like I have 50 state dimensions, and 1 counting dimension
I can see each of the state dimensions, would know by definition which state they are.
So they could clearly have a "state_callback' which would be 1 for cities in the state, and 0 for cities not in the state

But the counting dimension ... not sure how this would be setup ... as how would a single dimension know which "state' it was in for a given node

Sorry ... still wrapping my head around dimensions I think

Thanks
Evan

**Reply by ra-esmith** [reply]
So if I just have a dimension for each state, so 50 of them
And the max capacity is set to 1 for this dimension

And my callback returns 1 for all cities in state, and 0 for all cities out of state.


Would that not do what I am seeking?  
I am dropping locations already with a penalty callback (to handle a limit on the number of cities overall).


I guess I will try and explore this a bit more ... and learn about dimensions

**Comment by jmarca** [comment]
One idea off the top of my head is that every state is a separate dimension. Each city provides a constant value of 1 in its containing state, and zero in all other states. Each vehicle has a max capacity of one in each state dimension. 

James

**Reply by ra-esmith** [reply]
so yes a state dimension per state works just fine.  so will think now about why ... I am going to also I think try this again with a solver and constraint and see if I can get that to work.  as a good exercise as I learn this.

thanks!!
